Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -0,0 +1,192 @@
+
+    var thumbWidth: CGFloat { return CGFloat(width / numCols) }
+    var thumbHeight: CGFloat { return thumbWidth * aspectRatio }
+    var count: Int {
+        if let dataSource = self.collectionView?.dataSource as? TopSitesDataSource {
+            return dataSource.data.count
+        }
+        return 0
+    }
+    var width: CGFloat { return self.collectionView?.frame.width ?? 0 }
+
+    let ToolbarHeight: CGFloat = 44
+    let StatusBarHeight: CGFloat = 20
+    let RowHeight: CGFloat = 70
+
+    var attrCache = [Int: UICollectionViewLayoutAttributes]()
+
+    var thumbSecHeight: CGFloat { return width / numCols * numRows }
+    private func getIndexAtPosition(y: CGFloat) -> Int {
+        if y < thumbSecHeight {
+            let row = Int(y / thumbHeight)
+            return min(count-1, max(0, row * Int(numCols)))
+        }
+        return min(count-1, max(0, Int((y - thumbSecHeight) / RowHeight + numRows * numCols)))
+    }
+
+    override func collectionViewContentSize() -> CGSize {
+        let c = CGFloat(count)
+        let offset: CGFloat = ToolbarHeight + StatusBarHeight + CGFloat(ContainerHeight)
+
+        if c <= numRows * numCols {
+            let row = floor(Double(c / numCols))
+        	return CGSize(width: width, height: CGFloat(row) * thumbHeight + offset)
+        }
+
+        let h = (c - numRows * numCols) * RowHeight
+        return CGSize(width: width, height: numRows * thumbHeight + h + offset)
+    }
+
+    override func layoutAttributesForElementsInRect(rect: CGRect) -> [AnyObject]? {
+        let start = getIndexAtPosition(rect.origin.y)
+        let end = getIndexAtPosition(rect.origin.y + rect.height)
+
+        var attrs = [UICollectionViewLayoutAttributes]()
+        if start == -1 || end == -1 {
+            return attrs
+        }
+
+        for i in start...end {
+            let indexPath = NSIndexPath(forItem: i, inSection: 0)
+            let attr = layoutAttributesForItemAtIndexPath(indexPath)
+            attrs.append(attr)
+
+            if CGFloat(i) >= (numRows * numCols)-1 {
+                let decoration = layoutAttributesForDecorationViewOfKind(SeperatorKind, atIndexPath: indexPath)
+                attrs.append(decoration)
+            }
+        }
+        return attrs
+    }
+
+    override func layoutAttributesForDecorationViewOfKind(elementKind: String, atIndexPath indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes! {
+        let decoration = UICollectionViewLayoutAttributes(forDecorationViewOfKind: elementKind, withIndexPath: indexPath)
+
+        let h = ((CGFloat(indexPath.item + 1)) - numRows * numCols) * RowHeight
+        decoration.frame = CGRect(x: 0,
+            y: CGFloat(thumbHeight * numRows + h),
+            width: width,
+            height: 1)
+        return decoration
+    }
+
+    override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes! {
+        if let cached = attrCache[indexPath.item] {
```

Comment:
Do we really need to cache this already? It makes me nervous because I don't see code to evict items from the cache. Which probably means that at one point this cache will return invalid items. For example when the collection view is reloaded. Or when the device orientation is changed and the attributed need to be recalculated.

I would get rid of this cache until we actually see a performance problem. Also, knowing Apple, they probably cache the result of this delegate method already. And probably have all the logic in the collectionview to know when it needs to be calculated again.